libp2p/muxers/mplex/lpchannel.nim

```diff
@@ -128,52 +126,73 @@ method closed*(s: LPChannel): bool =
   ## header of the file
   s.closedLocal
 
+method pushTo*(s: LPChannel, data: seq[byte]) {.async.} =
+  if s.isReset:
+    raise newLPStreamClosedError() # Terminate mplex loop
+
+  try:
+    s.pushing = true
+    await procCall BufferStream(s).pushTo(data)
+  finally:
+    s.pushing = false
+
 method reset*(s: LPChannel) {.base, async, gcsafe.} =
   if s.closedLocal and s.isEof:
     trace "channel already closed or reset", s
     return
 
   trace "Resetting channel", s
 
-  asyncSpawn s.resetMessage()
+  # First, make sure any new calls to `readOnce` and `pushTo` will fail - there
+  # may already be such calls in the event queue
+  s.isEof = true
+  s.isReset = true
 
-  try:
-    # drain the buffer before closing
-    await s.drainBuffer()
-    await procCall BufferStream(s).close()
+  s.readBuf = StreamSeq()
 
-    s.isEof = true
-    s.closedLocal = true
+  s.closedLocal = true
 
-  except CancelledError as exc:
-    raise exc
-  except CatchableError as exc:
-    trace "Exception in reset", exc = exc.msg, s
+  asyncSpawn s.resetMessage()
+
+  # This should wake up any readers by pushing an EOF marker at least
+  await procCall BufferStream(s).close() # noraises, nocancels
+
+  if s.pushing:
+    # When data is being pushed, there will be two items competing for the
```

Comment:
this sounds like it would reintroduce the issue that `drainBuffer` was trying to solve?

Reply:
meaning that, in some cases, data that could have been read will be lost?

Reply:
it's slightly different: `drainBuffer` solved the regular controlled "close" case where the shutdown of the channel was controlled (one node signals "closed" - the other end reads all data and also signals closed etc) - `reset` is more for the extraordinary cases where something is already wrong - at that point, it's hard to reason about lost data, but generally yes, on `reset`, data will be lost - this is why it's important to be careful and only use it in extraordinary situations (timeout, underlying connection dead etc)

Reply:
Yes, that's why it was originally not used in resets and only added after you pointed out an issue in NBC, but I wonder if this issue in particular actually points to something in NBC proper, rather than the reset itself? 

In any case, I agree `reset` is destructive and should only happen in cases where we don't know what the state of the underlying socket is.

Reply:
well, initially it was not used anywhere and that was the issue - the eof flag was set even when there was data to process and no draining would happen - this branch takes another approach: it puts EOF on the queue together with other data - this ensures it's processed it the correct order.

`reset` upsets the order, and indeed, data gets lost - but then again, `reset` is only available internally - NBC doesn't have access to it, so it can't do anything wrong even if it wanted to.

there are still issues lurking here: there are 3 flags: `isClosed`, `isEof` and `closedLocal` for 2 properties (write-eof and read-eof) causing overlap - `BufferedStream` in particular doesn't understand or know about `closedLocal` which causes confusion and potentially race conditions - to properly implement this, one would have to support half-closed starting in `LPStream` and work up through the inheritance structure.

Reply:
the setup before this branch was particularly vulnerable to issues because it set flags on the global state, then called await - this would open it up for reordering issues - this is why the new approach uses the queue for EOF, to avoid these kinds of races

Reply:
> there are still issues lurking here: there are 3 flags: isClosed, isEof and closedLocal for 2 properties (write-eof and read-eof) causing overlap - BufferedStream in particular doesn't understand or know about closedLocal which causes confusion and potentially race conditions - to properly implement this, one would have to support half-closed starting in LPStream and work up through the inheritance structure.

The flags `isClosed`, `isEof` and `closedLocal` are supposed to be private, to access the state of the stream, the `closed` and `atEof` properties should be used. This was intended to allow propagating state across inheritance boundaries and for greater extensibility.

We can of course move the half-closed functionality all the way down to `LPStream`.